payment/datastore.go

```diff
@@ -0,0 +1,312 @@
+		SELECT SUM(amount) as sum
+		FROM transactions
+		WHERE order_id = $1 AND status = 'completed'
+	`, orderID)
+
+	return sum, err
+}
+
+// InsertIssuer inserts the given issuer
+func (pg *Postgres) InsertIssuer(issuer *Issuer) (*Issuer, error) {
+	statement := `
+	insert into order_cred_issuers (merchant_id, public_key)
+	values ($1, $2)
+	returning *`
+	issuers := []Issuer{}
+	err := pg.DB.Select(&issuers, statement, issuer.MerchantID, issuer.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(issuers) != 1 {
+		return nil, errors.New("Unexpected number of issuers returned")
+	}
+
+	return &issuers[0], nil
+}
+
+// GetIssuer retrieves the given issuer
+func (pg *Postgres) GetIssuer(merchantID string) (*Issuer, error) {
+	statement := "select * from order_cred_issuers where merchant_id = $1"
+	issuers := []Issuer{}
+	err := pg.DB.Select(&issuers, statement, merchantID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(issuers) > 0 {
+		return &issuers[0], nil
+	}
+
+	return nil, nil
+}
+
+// GetIssuerByPublicKey or return an error
+func (pg *Postgres) GetIssuerByPublicKey(publicKey string) (*Issuer, error) {
+	statement := "select * from order_cred_issuers where public_key = $1"
+	issuers := []Issuer{}
+	err := pg.DB.Select(&issuers, statement, publicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(issuers) > 0 {
+		return &issuers[0], nil
+	}
+
+	return nil, nil
+}
+
+// InsertOrderCreds inserts the given order creds
+func (pg *Postgres) InsertOrderCreds(creds *OrderCreds) error {
+	statement := `
+	insert into order_creds (order_id, issuer_id, blinded_creds)
+	values ($1, $2, $3)`
+	_, err := pg.DB.Exec(statement, creds.ID, creds.IssuerID, creds.BlindedCreds)
+	return err
+}
+
+// GetOrderCreds returns the order credentials for a OrderID
+func (pg *Postgres) GetOrderCreds(orderID uuid.UUID) (*[]OrderCreds, error) {
+	orderCreds := []OrderCreds{}
+	err := pg.DB.Select(&orderCreds, "select * from order_creds where order_id = $1", orderID)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(orderCreds) > 0 {
+		return &orderCreds, nil
+	}
+
+	return nil, nil
+}
+
+// RunNextOrderJob to sign order credentials if there is a order waiting, returning true if a job was attempted
+func (pg *Postgres) RunNextOrderJob(ctx context.Context, worker OrderWorker) (bool, error) {
+	tx, err := pg.DB.Beginx()
+	attempted := false
+	if err != nil {
+		return attempted, err
+	}
+
+	type SigningJob struct {
+		Issuer
+		OrderID      uuid.UUID                 `db:"order_id"`
+		BlindedCreds jsonutils.JSONStringArray `db:"blinded_creds"`
+	}
+
+	statement := `
+select
+	order_cred_issuers.*,
+	order_cred.order_id,
+	order_cred.blinded_creds
+from
+	(select *
+	from order_creds
+	where batch_proof is null
+	for update skip locked
+	limit 1
+) order_cred
+inner join order_cred_issuers
+on order_cred.issuer_id = order_cred_issuers.id`
+
+	jobs := []SigningJob{}
+	err = tx.Select(&jobs, statement)
+	if err != nil {
+		_ = tx.Rollback()
```

Comment:
we can probably just use defer instead of writing tx.Rollback every time we have a return

Reply:
Hope this is okay! Going to keep this how it is based on other uses in the repo 😅 I think it was mentioned to me that defer doesn't always clean up everything or work as intended?